[EM-599] Add support to truncate down to bytesize

jwsf opened the pull request:
**Description**

The main purpose of this PR is to add support for truncating an HTML document by bytesize instead of by character. It also adds some commonly-used code-cleanliness tools.
- [x] Link to TravisCI.org
- [x] Add Rubocop
- [x] Add Flay
- [x] Add bundle:audit
- [x] Add Brakeman
- [x] Check performance is proportional to truncation size (not input size)
- [x] Check memory usage is proportional to truncation size (not input size)
- [x] Strip out features we don't need - stripping elements, stripping attributes, altering where the tail goes, not including elements and/or attributes in bytesize count, counting by character, always include comments (they contain styles)
- [x] Output is HTML-encoded, HTML entities don't get truncated part-way thorugh
- [x] Support cdata (for css inside comments inside <style> elements
- [x] Comments and CData are not truncated - they are included in their entirety or omitted
- [x] Allow HTML entities to be used in tail

Here is an example of the truncation (from [here](https://github.com/zendesk/truncato/blob/08a9408d0318b4a65158aad779a28637e8864ee8/spec/fixtures/real_world_example.html) & [here](https://github.com/zendesk/truncato/blob/08a9408d0318b4a65158aad779a28637e8864ee8/spec/fixtures/real_world_example_truncated.html))

![html truncation 2](https://cloud.githubusercontent.com/assets/2756561/19955817/5c7c6046-a142-11e6-85aa-fb99bb780bbf.png)

jwsf committed:
Remove unused code

jwsf committed:
Add support for truncating by bitesize

Also fixed a load of edge conditions…

jwsf committed:
Refactor new parameter name

jwsf committed:
Rubocop compliance

jwsf committed:
Gem bump

jwsf committed:
Support multi-byte tails

jwsf committed:
Validate result is always valid HTML and re-factor tests

jwsf committed:
Test speed doesn't get slower with larger documents

Also had to bump a load of versions to get things working with
spec-benchmark

alduethadyn commented on spec/spec_helper.rb:
```diff
@@ -14,4 +15,13 @@
+    # Fires exception
+    Nokogiri::HTML("<html><body>#{actual}</body></html>", &:strict)
+    true
+  end
```
❤️

alduethadyn replied:
I want to do something like this in classic

jwsf committed:
Tweak .travis.yml

jwsf committed:
Travis and code quality goodness

jwsf committed:
Remove string-scrub

It is part of Ruby now…

jwsf committed:
Add test to check memory usage

mariozaizar commented on .gitignore:
```diff
@@ -21,3 +21,5 @@ doc/
 
 
 
+
```
🎨  Remove extra line spaces.

mariozaizar commented on .rubocop.yml:
```diff
@@ -0,0 +1,182 @@
+    - 'Rakefile'
+  Exclude:
+    - 'gemfiles/**/*'
+    - 'vendor/cache/**/*'
```
:art: Add EOF line.

mariozaizar commented on .ruby-version:
```diff
@@ -1 +1 @@
-2.0.0-p195
\ No newline at end of file
+2.2.3
```
Nice.

mariozaizar commented on .travis.yml:
```diff
@@ -1,4 +1,16 @@
-  - 2.0.0
+  - 2.2.3
+  - 2.2.4
+
```
:art: Remove line.

mariozaizar commented on .travis.yml:
```diff
@@ -1,4 +1,16 @@
+      script: 'bundle exec rubocop'
+      script: 'bundle exec flay'
+  exclude:
+    - rvm: 2.2.4
```
Why we add 2.2.4 but then exclude? Are the specs breaking with this version?

jwsf replied:
It just evolved that way. Removing.

mariozaizar commented on Gemfile:
```diff
@@ -10,10 +9,12 @@ gemspec
 end
 
-group :benchrmark do
+group :benchmark do
```
❤️

mariozaizar commented on Rakefile:
```diff
@@ -28,5 +25,10 @@ Rake::TestTask.new(:test) do |t|
+RuboCop::RakeTask.new
+
+require 'bundler/audit/task'
+Bundler::Audit::Task.new
```
Did you got any outstanding issue?

jwsf replied:
Nope :-)

mariozaizar reviewed (commented):
:eyes:

jwsf committed:
Code review tidy

rmartz00 commented on lib/truncato/truncated_sax_document.rb:
```diff
@@ -107,76 +163,89 @@ def opening_tag name, attributes
-  def overriden_tag_length
-    @count_tags ? nil : 0
-  end
+  def overriden_tag_length(tag_name, rendered_tag_with_attributes)
```
"overridden"

jwsf committed:
Typo

rmartz00 commented on lib/truncato/truncated_sax_document.rb:
```diff
@@ -2,41 +2,79 @@
+      # We're counting characters, so don't worry how many bytes it takes to reposent this in the string
+      char_or_byte_count(string_to_append)
+    end
+    append_to_truncated_string encoded_string_to_append, length
```
this seems like it would be different behavior than the original, which would pass `string_to_append.length` even though the string itself was encoded..

alduethadyn replied:
Also, parens

rmartz00 commented on lib/truncato/truncated_sax_document.rb:
```diff
@@ -46,7 +84,9 @@ def end_element name
     unless single_tag_element? name
       @closing_tags.pop
-      append_to_truncated_string closing_tag(name), overriden_tag_length
+      # Don't count the length when closing a tag - it was accomodated when
```
"accommodated"

rmartz00 commented on lib/truncato/truncated_sax_document.rb:
```diff
@@ -56,49 +96,65 @@ def end_document
+
+  def estimated_length_with_tail
+    # byebug
+    @estimated_length + (@count_tail && @something_has_been_truncated ? tail_length : 0)
```
parens around `@count_tail && @ something_has_been_truncated`? found it confusing upon first reading..also remove byebug comment above

rmartz00 commented on lib/truncato/truncated_sax_document.rb:
```diff
@@ -56,49 +96,65 @@ def end_document
   def process_comment(string)
-    remaining_length = max_length - @estimated_length - 1
-    string_to_append = comment_tag(string).length > remaining_length ? truncate_comment(comment_tag(string), remaining_length) : comment_tag(string)
+    string_to_append = if char_or_byte_count(comment_tag(string)) > remaining_length
```
maybe pull out `comment_tag(string)` into a variable?

jwsf committed:
Make sure all tests run on Travis

jwsf committed:
Code-review tidy

jwsf committed:
Remove unused features and fix some more bugs

jwsf committed:
Get things working with real-world exmple

+ Fix HTML entities
+ Fix comment css in style blocks (data support)
+ Add comments back in (they have CSS)
+ Truncated output is HTML-encoded
+ Comments and data blocks are not truncated - they are either added in
their entirety or committed

jwsf committed:
Extra comments

jwsf committed:
Update Brakeman to 3.4.1

alduethadyn commented on benchmark/truncato/vendor/peppercorn_adapter.rb:
```diff
@@ -1,9 +1,8 @@
 # Adapter for comparing https://github.com/nono/HTML-Truncator
 module Truncato
```
I assume the module name will be updated?

rmartz00 commented on lib/truncato/truncated_sax_document.rb:
```diff
@@ -2,42 +2,92 @@
+    end
+
+    # Save the tag so we can push it on at the end
     @closing_tags.push name unless single_tag_element? name
```
Seems like this makes an assumption that the `VOID_TAGS` would never be seen with a closing tag, but is that really true? Isn't it possible that some client out there somewhere closes one of those tags even though it does not need to? Is this enforced some other way? Or another way to ask the question is, what happens if it _does_ happen to have a corresponding closing tag but we are not pushing it here onto the stack?

jwsf replied:
@rmartz00 I added some test cases to show how Nokogiri is cleaning this up for us: https://github.com/zendesk/abbreviato/commit/3977c107fd131c65491d819b4d113ac5637b919f

jwsf committed:
Complete name change to Abbreviato

jwsf committed:
Add brake check gem

jwsf committed:
Update gems

jwsf committed:
Support HTML entities as tails

jwsf committed:
Some additional test cases

jwsf committed:
More test cases

rmartz00 commented on spec/fixtures/real_world_example.html:
```diff
@@ -0,0 +1,973 @@
+<br>
+<span style="font-size:10.0pt;font-family:&quot;Calibri&quot;,sans-serif;color:#004080">Gibb</span>
+<br>
+<span style="font-size:10.0pt;font-family:&quot;Calibri&quot;,sans-serif;color:#004080">513-383-5019</span>
```
Has this file been fully sanitized of PII?

jwsf replied:
Yes, all emails, names, corp names, phone numbers, addresses, etc replaced

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,198 @@
+
+  def attributes_to_string(attributes)
+    return "" if attributes.empty?
+    attributes_string = concatenate_attributes_declaration attributes
```
`concatenate_attributes_declaration(attributes).rstrip`?

jwsf replied:
@rmartz00 Done (and more)

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,198 @@
+  end
+
+  def check_max_length_reached
+    @max_length_reached = true if estimated_length_with_tail >= max_length
```
what's the reason to keep this as an instance variable rather than just doing the calculation each time it's needed?

jwsf replied:
@rmartz00 I inherited that. The code feels cleaner without it...

jwsf committed:
Tidy after code review

jwsf committed:
Remove use of @max_length_reached

Memorization vs. making the code clearer…

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,189 @@
+
+  def truncate_string(string)
+    @something_has_been_truncated = true
+    (string.byteslice(0, remaining_length) || '').scrub('') unless remaining_length < 0
```
does the check on `unless remaining_length < 0` (or `if remaining_length >= 0` as is used elsewhere) need to be wrapped around everything here including the `@something_has_been_truncated = true`? seems like you wouldn't want to set that unless you were actually truncating..

jwsf replied:
Actually, it is not needed at all anymore as the check is done higher up. I've tidied it up a bit.

https://github.com/zendesk/abbreviato/blob/jwilson/EM-599-bytesize/lib/abbreviato/truncated_sax_document.rb#L57

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,189 @@
+  end
+
+  def estimated_length_with_tail
+    @estimated_length + (@something_has_been_truncated ? tail.bytesize : 0)
```
I'm not sure I get why we only include the `tail`'s length here if something has been truncated - the name of the method says `with_tail` after all.. Shouldn't we always be looking at the estimated length including the tail if one has been requested?

rmartz00 replied:
[This line](https://github.com/zendesk/abbreviato/pull/1/files#diff-3c291160de3d24bf025ee43f9ca67804R60) in particular seems more confusing because of that.. I would expect the `tail`'s length to already be accounted for in the earlier check against `remaining_length` so we can just add it if needed.

jwsf replied:
Ah, I see what you mean. Logic updated.

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,189 @@
+    else
+      encoded_string
+    end
+    append_to_truncated_string string_to_append, string_to_append.bytesize
```
looks like `string_to_append.bytesize` doesn't _need_ to be passed in here..

jwsf replied:
Good catch

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,189 @@
+      return
+    end
+
+    return if max_length_reached || ignorable_tag?(name) || ignore_mode?
```
`|| ignore_mode?` not needed I think..

jwsf replied:
@rmartz00 Correct!

jwsf committed:
More code tidy-up after code review

jwsf committed:
Remove extraneous check

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,182 @@
+  # This method is called when the parser encounters characters between tags
+  def characters(decoded_string)
+    if max_length_reached || ignore_mode?
+      @something_has_been_truncated = max_length_reached
```
I'm not sure I understand the need for `@something_has_been_truncated` either.. it seems to only be used in a few places within this one function..the only place we actually check its value is [here](https://github.com/zendesk/abbreviato/pull/1/files#diff-3c291160de3d24bf025ee43f9ca67804R61) but we set it to `true` 2 lines before so it's always true by the time we get there..

jwsf replied:
@rmartz00 Agreed - it's a holdover from earlier (more complex) versions

jwsf replied:
Removed.

jwsf committed:
Remove use of @something_has_been_truncated

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,177 @@
+    # Abort if there is not enough space to add the combined opening tag and (potentially) the closing tag
+    length_of_tags = overridden_tag_length(name, string_to_add)
+    if length_of_tags > remaining_length
+      enter_ignored_level
```
do we still want to `enter_ignored_level` even if it's a single-tag element?

jwsf replied:
@rmartz00 Fixed

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,177 @@
+  end
+
+  def remaining_length
+    max_length - @estimated_length
```
why would we not account for `tail.length` here instead of elsewhere?

jwsf replied:
`@estimated_length` now includes the length of the tail if it has been added. The only accommodation for the tail is to leave space for it when truncating the string. here: https://github.com/zendesk/abbreviato/pull/1/files#diff-3c291160de3d24bf025ee43f9ca67804R149 @rmartz00

rmartz00 replied:
okay, I guess I was thinking that we should be leaving the buffer for the tail in case it is needed - otherwise if `remaining_length` is exactly the tail's length, we might end up thinking we can add more content but then we can't. [This line](https://github.com/zendesk/abbreviato/pull/1/files#diff-3c291160de3d24bf025ee43f9ca67804R62) in particular seems odd - don't we always want to add the tail if it's been requested? Doesn't the current implementation mean we could end up not adding the tail in some cases?

jwsf replied:
@rmartz00 After some discussion, decided to leave it this way so we only add a tail if we truncate some text, and not if we drop entire HTML blocks.

rmartz00 replied:
so if there are cases where we don't include the tail (dropping entire HTML blocks, or the small edge case above where we don't add the tail if the encoded HTML ends up taking more space than can accommodate the tail), how does the caller know if something has been truncated? It would need to in order for us to know to show the appropriate UI, right? I thought we were going to use a tail being present in the stored text/HTML to signify this..

jwsf replied:
I've now added an additional return parameter which indicates whether the HTML was actually truncated.

alduethadyn commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,177 @@
+
+  # These don't have to be closed (which also impacts ongoing length calculations)
+  # http://www.456bereastreet.com/archive/201005/void_empty_elements_and_self-closing_start_tags_in_html/
+  VOID_TAGS = %w[area base br col command hr img input keygen link meta param source].freeze
```
we should compare this with what zendesk comment markup has in terms of void tags...

alduethadyn replied:
Mainly I am thinking `wbr`

alduethadyn replied:
Which has weird rules.

alduethadyn commented on spec/abbreviato/abbreviato_spec.rb:
```diff
@@ -0,0 +1,265 @@
+
+  describe "normal strings" do
+    it_should_truncate "no html text with longer length",
+      with: {max_length: 99},
```
nit: `{max_length: 99}` should be: `{ max_length: 99 }`

alduethadyn commented on spec/abbreviato/abbreviato_spec.rb:
```diff
@@ -0,0 +1,265 @@
+  let(:html_10Mb_doc) { File.read('spec/fixtures/html_1Mb.html') }
+  let(:bench) { Benchmark::Perf::ExecutionTime.new(samples: 10) }
+
+  it "speed is proportional to length of truncated string, not input" do
```
👍

jwsf committed:
Fix situation where a long tag is followed by a short tag

So that it correctly includes the shorter tag.

And a bunch of other random tests

jwsf committed:
Tweak formatting in tests

jwsf committed:
Fix parenthesis

jwsf committed:
Add checks to ensure void tags behave as expected (nokogiri treats them correctly)

jwsf committed:
Add support for <wbr> element

jwsf committed:
Fix broken test

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,182 @@
+    string_to_append = if encoded_string.bytesize > remaining_length
+      # This is the line which prevents HTML entities getting truncated - treat them as a single char
+      str = @html_coder.encode(truncate_string(decoded_string), :named)
+      str << tail if remaining_length >= tail.bytesize
```
Shouldn't this be checking `remaining_length - str.bytesize >= tail.bytesize` since `remaining_length` doesn't account for the new (truncated) string yet?

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,182 @@
+    append_to_truncated_string(string_to_append)
+  end
+
+  # This method is called when the parser encounters an comment
```
"a comment"

jwsf committed:
Tweak logic for checking whether there is space for the tail to be added

jwsf committed:
Include return value on whether the string was truncated

jwsf committed:
Removed benchmark as no longer comparing apples-to-apples

Characters versus bytes

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -59,18 +61,22 @@ def characters(decoded_string)
       # This is the line which prevents HTML entities getting truncated - treat them as a single char
       str = @html_coder.encode(truncate_string(decoded_string), :named)
-      str << tail if remaining_length >= tail.bytesize
+      str << tail if remaining_length - str.bytesize >= tail.bytesize
```
is it ever possible for `tail` or `str` to be `nil`?

jwsf replied:
`tail` is now protected [here](https://github.com/zendesk/abbreviato/pull/1/files#diff-3c291160de3d24bf025ee43f9ca67804R21) and I think it unlikely that str will be nil, but additional check added [here](https://github.com/zendesk/abbreviato/pull/1/files#diff-3c291160de3d24bf025ee43f9ca67804R63) @rmartz00

rmartz00 commented on spec/support/spec_helpers/abbreviato_macros.rb:
```diff
@@ -1,12 +1,12 @@
 module AbbreviatoMacros
-  def it_should_truncate(example_description, options)
+  def it_should_truncate(example_description, should_truncate, options)
```
confused by the name of the new parameter v. the name of the function..

jwsf replied:
@rmartz00 Hmm... this 'macro' is used to test both situations - when the gem is expected to truncate the supplied string and also when it is not (because the supplied string is short). Thoughts on a better name for the method? `it_truncates_if_needed`?

rmartz00 replied:
maybe rename this to `test_truncate` then 2 other helper methods called `it_truncates` and `it_does_not_truncate` that pass in the appropriate value for `should_truncate` to `test_truncate` (and also modify the description used in `test_truncate` accordingly). then wherever `it_should_truncate` is being called, modify to use the more specific method and not have to pass the boolean param each time.

jwsf replied:
@rmartz00 Done

jwsf committed:
Protect against nil parameters

jwsf committed:
Missing space

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,194 @@
+    string_to_append = if encoded_string.bytesize > remaining_length
+      # This is the line which prevents HTML entities getting truncated - treat them as a single char
+      str = @html_coder.encode(truncate_string(decoded_string), :named)
+      str << tail if remaining_length - str.bytesize >= tail.bytesize
```
so it isn't possible for `str` to be `nil`?

jwsf replied:
Hmm... I've protected against the only way I've been able to generate `nil` here already but it won't hurt to have additional checks. I'll make that change...

jwsf replied:
@rmartz00 Done

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,194 @@
+      exit_ignored_level unless single_tag_element?(name)
+      return
+    end
+    return if max_length_reached || ignorable_tag?(name)
```
if `max_length_reached`, does `@truncated` need to be set to true?

jwsf replied:
No, this is (now) taken care of in `start_element` @rmartz00

rmartz00 replied:
Couldn't there be content added to the truncated string in `start_element` though that could push it to then be at `max_length_reached` by the time `characters` is called? Otherwise why does the `max_length_reached` check need to be in `characters` then? Seems like there needs to be an additional `@truncated = true` in `characters` if `max_length_reached` (or DRY it up in `max_length_reached` instead).

jwsf replied:
Oooo you're good! Fixed and added a test case. I feel like I had all this `@truncated` code in previously...

jwsf replied:
I'm resisting the temptation to set `@truncated` in `max_length_reached` as it feels like I'm adding an unexpected side-effect. It is also being explicitly set in other places.

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,194 @@
+require 'htmlentities'
+
+class TruncatedSaxDocument < Nokogiri::XML::SAX::Document
+  IGNORABLE_TAGS = %w[html head body].freeze
```
why are these ignorable? do they get generated by Nokogiri elsewhere? do we account for their bytesizes?

alduethadyn replied:
I think it is because they are non-truncateable

rmartz00 replied:
but then we don't seem to add them back anywhere nor account for their sizes in the overall truncated string length.

jwsf replied:
At the moment, these elements get stripped. Depending on the nature of the HTML documents/fragments we're going to be truncating, this part will need to be re-visited. Do we know if the email parts will still have these elements by the time they need to be truncated? @alduethadyn @rmartz00

rmartz00 replied:
I think we need to look at some of the current examples to answer this question..

alduethadyn replied:
We end up without the body (just the inner HTML) at the end, so I think this is safe/ok... see `Content#html`

jwsf replied:
@rmartz00 I'm planning to test this within the context of our larger codebase once I get the gem there - that way I can force parsing of a bunch of our fixtures and check the results in context of all the other sanitization and parsing which is going on.

rmartz00 commented on lib/abbreviato/abbreviato.rb:
```diff
@@ -0,0 +1,22 @@
+      tail: "..."
+  }.freeze
+
+  # Truncates the source XML string and returns the truncated XML. It will keep a valid XML structure
```
should this be modified to include info about the return of the `.truncated` field?

jwsf replied:
@rmartz00 Done

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,194 @@
+
+  # This method is called when the parser encounters an open tag
+  def start_element(name, attributes)
+    return if max_length_reached || ignorable_tag?(name)
```
same question here about whether `@truncated` needs to be true if `max_length_reached` is true

jwsf replied:
Good catch. Fixed and added tests. @rmartz00

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,194 @@
+  end
+
+  def enter_ignored_level
+    @truncated = true
```
seems like we need to move the checks on `single_tag_element?` that accompany the calls to this function to move into the function itself so that we still get the `@truncated = true` for the single tag elements

jwsf replied:
Another good catch. However, the only place I've needed to add it is [here](https://github.com/zendesk/abbreviato/pull/1/files#diff-3c291160de3d24bf025ee43f9ca67804R48).  @rmartz00

rmartz00 replied:
right, I think you could DRY up the `single_tag_element?` checks accompanying the `enter_ignored_level` calls though by passing in the `name` to `enter_ignored_level` and just doing the `single_tag_element?` check inside..

jwsf committed:
Additional protection against nil string

jwsf committed:
Rename test macros for clarity

jwsf committed:
Update comments

jwsf committed:
Fix edge case where empty tags fit exactly

jwsf committed:
Fix edge case with void tags

jwsf committed:
Added dryness

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,197 @@
+  end
+
+  def enter_ignored_level(name)
+    @ignored_levels += 1 unless single_tag_element?(name)
```
wasn't there a `@truncated = true` here before?

jwsf replied:
@rmartz00 There was, but it is only really needed when we enter the ignored_level which happens [here](https://github.com/zendesk/abbreviato/blob/jwilson/EM-599-bytesize/lib/abbreviato/truncated_sax_document.rb#L48)

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,197 @@
+  end
+
+  def exit_ignored_level
+    @ignored_levels -= 1
```
same DRYness here with `single_tag_element?` check for consistency? (even though it is only called in one place)

rmartz00 commented on spec/abbreviato/abbreviato_spec.rb:
```diff
@@ -0,0 +1,330 @@
+      source: "<o:p>&nbsp;</o:p>",
+      expected: "<o:p>&nbsp;</o:p>"
+
+    it_truncates "not truncate html entities (all or nothing)",
```
this description contradicts the `it_truncates`?

jwsf replied:
@rmatrz Technically doesn't contradict, but I've improved the description to make it clearer what's being tested.

jwsf committed:
Added consistency

jwsf committed:
Better test description

jwsf committed:
Another edge case - tags which fit perfectly within inner content

jwsf committed:
max_length_reached -> mac_length_reached?

jwsf commented:
@rmartz00 Another look?

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,200 @@
+      exit_ignored_level(name)
+      return
+    end
+    return if max_length_reached? || ignorable_tag?(name)
```
need `@truncated = true` here?

jwsf replied:
I've not been able to concoct a test case to prove it is needed. you think it worth adding regardless? @rmartz00

rmartz00 replied:
is that because it's actually impossible for `max_length_reached?` to be true here?

jwsf replied:
From what I recollect trying to break this last week,` max_length_reached` can be true, but `@truncated` is also already set.

rmartz00 replied:
It seems like this would be the case when you've already accounted for the closing tag in the `@estimated_length` and it fits exactly in the truncated string. However, since we return early here, you would never get to add that closing tag..

rmartz00 replied:
so maybe `max_length_reached?` should not actually be checked here..

jwsf replied:
No, there are situations where `max_length_reached` is `true`, but `@truncated` is _already_ set to `true`.

rmartz00 replied:
Okay regardless of whether `@truncated` would already be true, what about the case I mentioned earlier where `max_length_reached?` is true because the `estimated_length` equals the `max_length` exactly because the `estimated_length` already includes the end tag length (from when `start_element` was called) - we wouldn't actually get to add that closing tag here because we'd be returning early, right?

rmartz00 replied:
oh I guess maybe there is an assumption that `end_document` will take care of them..since it doesn't check length info at all.

jwsf replied:
@rmartz I'm reluctant to change the code here without a broken test case to solve. I have a bunch of them [here](https://github.com/zendesk/abbreviato/blob/jwilson/EM-599-bytesize/spec/abbreviato/abbreviato_spec.rb#L228). Can you think of a scenario I've missed which would break because of this concern?? I can't come up with one. Also, a lot of the other test cases have very exact max lengths and the closing tags get added as the document closes.

rmartz00 replied:
> have very exact max lengths and the closing tags get added as the document closes.

Yes, this is what I realized in my last comment.. I think it's fine but maybe a comment here that that would be the case (closing tags will get added as the document closes if max_length_reached? is true)

jwsf replied:
Done!

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,200 @@
+  end
+
+  def end_document
+    @closing_tags.reverse.each { |name| append_to_truncated_string(closing_tag(name)) }
```
use `reverse_each` instead?

jwsf replied:
@rmartz00 Done

jwsf committed:
Use reverse_each

jwsf committed:
Add option to parse HTML documents as well as fragments

alduethadyn commented on spec/abbreviato/abbreviato_spec.rb:
```diff
@@ -0,0 +1,373 @@
+      expected: "<p>hello...</p>"
+  end
+
+  describe "document mode" do
```
Are test cases needed to handle other charsets with encoded characters? Like ISO-8859-1: Å, å, Ä, ä, Ö, ö etc?

jwsf committed:
Tweak how nil source is handled

jwsf committed:
Update gems and make more use of brakecheck

jwsf committed:
Change default tail to &hellip;

jwsf committed:
Tests for non-english characters

jwsf committed:
Add brakecheck to brakecheck-check!

rmartz00 commented on lib/abbreviato/truncated_sax_document.rb:
```diff
@@ -0,0 +1,201 @@
+  end
+
+  def end_document
+    @closing_tags.reverse_each { |name| append_to_truncated_string(closing_tag(name)) }
```
don't we need to pass in 0 for the length here?

jwsf replied:
It doesn't really matter when we're closing the document as `@estimated_length` is not going to be looked at again, but it makes the code clearer so I've made the change.

rmartz00 commented on lib/abbreviato/abbreviato.rb:
```diff
@@ -0,0 +1,24 @@
+  # @param [Hash] user_options truncation options
+  # @option user_options [Integer] :max_length Maximum length
+  # @option user_options [String] :tail text to append when the truncation happens
+  # @return [[String] the truncated string, [boolean] whether the string was truncated]
```
need documentation on `fragment` mode?

jwsf committed:
Improve documentation

jwsf committed:
Comment how end tags are handled

rmartz00 commented:
did the tests get lost? I thought there were tests here before that use the new `it_truncates` type helpers..otherwise this is looking good..

jwsf commented:
Which tests exactly? There are a bunch [here](https://github.com/zendesk/abbreviato/blob/jwilson/EM-599-bytesize/spec/abbreviato/abbreviato_spec.rb#L228) which looks at edge cases around tag/truncation length. @rmartz00

rmartz00 commented:
ah that was the file I was looking for, thanks! for some reason github wasn't showing me the diff of that one by default (I had to click "Load diff" :/ )
